BUILD_DETAIL.md

```diff
@@ -0,0 +1,144 @@
+Our commitment to [SemVer](htp://semver.org) requires that we explicitly
+declare our public API, and our build uses YARD to ensure that every
+class, module and method has either been labeled `@private` or has at
+least some level of documentation. For new APIs, this forces us to make
+an intentional decision about whether or not it should be part of
+RSpec's public API or not.
+
+To run the YARD documentation coverage check, run:
+
+```
+$ bundle exec yard stats --list-undoc
+
+# or, if you installed your bundle with `--standalone --binstubs`:
+
+$ bin/yard stats --list-undoc
+```
+
+We also want to prevent YARD errors or warnings when actually generating
+the docs. To check for those, run:
+
+```
+$ bundle exec yard doc --no-cache
+
+# or, if you installed your bundle with `--standalone --binstubs`:
+
+$ bin/yard doc --no-cache
+```
+
+## Rubocop
+
+We use [Rubocop](https://github.com/bbatsov/rubocop) to enforce style conventions on the project so
+that the code has stylistic consistency throughout. Run with:
+
+```
+$ bundle exec rubocop lib
+
+# or, if you installed your bundle with `--standalone --binstubs`:
+
+$ bin/rubocop lib
+```
+
+Our Rubocop configuration is a work-in-progress, so if you get a failure
+due to a Rubocop default, feel free to ask about changing the
+configuration. Otherwise, you'll need to address the Rubocop failure,
+or, as a measure of last resort, by wrapping the offending code in
+comments like `# rubocop:disable SomeCheck` and `# rubocop:enable SomeCheck`.
+
+## Run spec files one-by-one
+
+A fast TDD cycle depends upon being able to run a single spec file,
+without the rest of the test suite. While rare, it's fairly easy to
+create a situation where a spec passes when the entire suite runs
+but fails when its individual file is run. To guard against this,
+our CI build runs each spec file individually, using a bit of bash like:
+
+```
+for file in `find spec -iname '*_spec.rb'`; do
+  echo "Running $file"
+  bin/rspec $file -b --format progress
+done
+```
+
+Since this step boots RSpec so many times, it runs much, much
+faster when we can avoid the overhead of bundler. This is a main reason our
+CI build installs the bundle with `--standalone --binstubs` and
+runs RSpec via `bin/rspec` rather than `bundle exec rspec`.
+
+## Running the spec suite for each of the other repos
+
+While each of the RSpec repos is an independent gem (generally designed
+to be usable on its own), there are interdependencies between the gems,
+and the specs for each tend to use features from the other gems. We
+don't want to merge a pull request for one repo that might break the
+build for another repo, so our CI build includes a spec that runs the
+spec suite of each of the _other_ project repos. Note that we only run
+the spec suite, not the full build, of the other projects, as the spec
+suite runs very quickly compared to the full build.
+
```

Comment:
They are not in sync yet. I opened #131 to update our Rake task to get them in sync.

This came up with talking to the group at Ruby for Good. I asked them to go look and see if they could figure out how to get started helping with RSpec. The unanimous response was that it was difficult because there wasn't information in the README.

They've opened PRs on the other repos (I've merged them in) adding a section - it's included in this PR too. Part of that update is including a list to the relative docs (setup, contributing and CoC files) at the project root. That's what I wanted to do here as well.